### PR TITLE
Add load_by_name implementation into wasi-nn plugin

### PR DESCRIPTION
**Pull request by CaptainVincent**
~~Pending wait for https://github.com/bytecodealliance/wasi-nn/pull/87 merged first.~~ Merged.

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

File: plugins/wasi_nn/wasinnfunc.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include "common/log.h"
 #include "wasinnenv.h"
 
+#include <iostream>
```
Comment (hydai) [inline plugins/wasi_nn/wasinnfunc.cpp:8]: Do we need this header?

File: plugins/wasi_nn/wasinnenv.h [hunk 2]
```diff
@@ -145,13 +146,51 @@ struct WasiNNEnvironment :
     FOR_EACH_BACKEND(EACH)
 #undef EACH
         std::monostate {
-  WasiNNEnvironment() noexcept {
-    NNGraph.reserve(16U);
```
Comment (hydai) [inline plugins/wasi_nn/wasinnenv.h:149]: Why does this section remove?

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/wasinnfunc.cpp (hunk 1 above)
Comment (CaptainVincent) [reply plugins/wasi_nn/wasinnfunc.cpp]: remove done.

File: plugins/wasi_nn/wasinnenv.h (hunk 2 above)
Comment (CaptainVincent) [reply plugins/wasi_nn/wasinnenv.h:149]: I put the constructor definition into .cpp file. I thought they meant the same thing.
https://github.com/CaptainVincent/WasmEdge/blob/0b528de8307b3032acc3796b45a45937d12f638f/plugins/wasi_nn/wasinnenv.cpp#L77

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/wasinnenv.cpp [hunk 3]
```diff
@@ -7,13 +7,87 @@
...
+    {"Tensorflow", Backend::Tensorflow},
+    {"PyTorch", Backend::PyTorch},
+    {"TensorflowLite", Backend::TensorflowLite},
+    {"Autodetect", Backend::Autodetect}};
+
+std::map<std::string, Device> DeviceMap = {
+    {"CPU", Device::CPU}, {"GPU", Device::GPU}, {"TPU", Device::TPU}};
+
+bool load(const std::string &path, std::vector<uint8_t> &data) {
+  std::ifstream file(path, std::ios::binary);
```
[the remark this answers is not in the document]
Comment (CaptainVincent) [reply plugins/wasi_nn/wasinnenv.cpp]: Replace done.

File: plugins/wasi_nn/wasinnenv.cpp [hunk 4]
```diff
@@ -7,13 +7,87 @@
...
+    {"ONNX", Backend::ONNX},
+    {"Tensorflow", Backend::Tensorflow},
+    {"PyTorch", Backend::PyTorch},
+    {"TensorflowLite", Backend::TensorflowLite},
+    {"Autodetect", Backend::Autodetect}};
+
+std::map<std::string, Device> DeviceMap = {
+    {"CPU", Device::CPU}, {"GPU", Device::GPU}, {"TPU", Device::TPU}};
+
+bool load(const std::string &path, std::vector<uint8_t> &data) {
```
[the remark this answers is not in the document]
Comment (CaptainVincent) [reply plugins/wasi_nn/wasinnenv.cpp]: The case correction is complete.

File: plugins/wasi_nn/wasinnenv.cpp [hunk 5]
```diff
@@ -7,13 +7,87 @@
...
+  file.read(reinterpret_cast<char *>(data.data()), fileSize);
+  file.close();
+  return true;
+}
+
+WasiNNEnvironment::WasiNNEnvironment() noexcept {
+  // Preload NN Models
+  for (const auto &M : NNModels.value()) {
+    std::istringstream ISS(M);
+    char Delimiter = ':';
```
[the remark this answers is not in the document]
Comment (CaptainVincent) [reply plugins/wasi_nn/wasinnenv.cpp]: Add const to Delimiter done.

File: plugins/wasi_nn/wasinnenv.cpp [hunk 6]
```diff
@@ -7,13 +7,87 @@
...
-namespace {
+namespace WASINN {
 
 Runtime::Instance::ModuleInstance *
 create(const Plugin::PluginModule::ModuleDescriptor *) noexcept {
   return new WasiNNModule;
 }
 
+std::map<std::string, Backend> BackendMap = {
+    {"OpenVINO", Backend::OpenVINO},
```
Comment (hydai) [inline plugins/wasi_nn/wasinnenv.cpp:18]: Use `string_literal`, like `"OpenVINO"sv`.

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/wasinnenv.cpp (hunk 6 above)
Comment (CaptainVincent) [reply plugins/wasi_nn/wasinnenv.cpp]: Done.

**Review by hydai — APPROVED** [review]

**Pull request merged**